libp2p/src/builder.rs

```diff
@@ -0,0 +1,1301 @@
+// TODO: Rename runtime to provider
+// TODO: Should we have a timeout on transport?
+// TODO: Be able to address `SwarmBuilder` configuration methods.
+// TODO: Consider moving with_relay after with_other_transport.
+// TODO: Consider making with_other_transport fallible.
+
+use libp2p_core::{muxing::StreamMuxerBox, Transport};
+use std::marker::PhantomData;
+
+pub struct SwarmBuilder<P, Phase> {
+    keypair: libp2p_identity::Keypair,
+    phantom: PhantomData<P>,
+    phase: Phase,
+}
+
+pub struct InitialPhase {}
+
+impl SwarmBuilder<NoProviderSpecified, InitialPhase> {
+    pub fn with_new_identity() -> SwarmBuilder<NoProviderSpecified, ProviderPhase> {
+        SwarmBuilder::with_existing_identity(libp2p_identity::Keypair::generate_ed25519())
+    }
+
+    pub fn with_existing_identity(
+        keypair: libp2p_identity::Keypair,
+    ) -> SwarmBuilder<NoProviderSpecified, ProviderPhase> {
+        SwarmBuilder {
+            keypair,
+            phantom: PhantomData,
+            phase: ProviderPhase {},
+        }
+    }
+}
+
+pub struct ProviderPhase {}
+
+impl SwarmBuilder<NoProviderSpecified, ProviderPhase> {
+    #[cfg(feature = "async-std")]
+    pub fn with_async_std(self) -> SwarmBuilder<AsyncStd, TcpPhase> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: TcpPhase {},
+        }
+    }
+
+    #[cfg(feature = "tokio")]
+    pub fn with_tokio(self) -> SwarmBuilder<AsyncStd, TcpPhase> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: TcpPhase {},
+        }
+    }
+}
+
+pub struct TcpPhase {}
+
+#[cfg(feature = "tcp")]
+impl<P> SwarmBuilder<P, TcpPhase> {
+    pub fn with_tcp(self) -> SwarmBuilder<P, TcpTlsPhase> {
+        self.with_tcp_config(Default::default())
+    }
+
+    pub fn with_tcp_config(self, config: libp2p_tcp::Config) -> SwarmBuilder<P, TcpTlsPhase> {
+        SwarmBuilder {
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: TcpTlsPhase { config },
+        }
+    }
+}
+
+impl<P> SwarmBuilder<P, TcpPhase> {
+    // TODO: This would allow one to build a faulty transport.
+    pub fn without_tcp(
+        self,
+    ) -> SwarmBuilder<P, RelayPhase<impl AuthenticatedMultiplexedTransport>> {
+        SwarmBuilder {
+            // TODO: Is this a good idea in a production environment? Unfortunately I don't know a
+            // way around it. One can not define two `with_relay` methods, one with a real transport
+            // using OrTransport, one with a fake transport discarding it right away.
+            keypair: self.keypair,
+            phantom: PhantomData,
+            phase: RelayPhase {
+                transport: libp2p_core::transport::dummy::DummyTransport::new(),
+            },
```

Comment:
> Why should it be legal to go to the relay phase without first saying `with_tcp` or `with_quic`?

`with_tcp`, `with_quic` or `with_other_transport`

Fair. Let me see what I can do. Thus far I have shied away from its complexity.

Reply:
I don't see a reasonable way around this. One could make `with_quic` (...) generic over the previous transport, where a previous `DummyTransport` would lead to it being dropped in `with_quic` and `build` to only be available when the final transport is not a `DummyTransport`. I don't think it is worth the complexity this would introduce.